### ISSUE TITLE
Remove `Rebroadcast`s from graphs

### ISSUE THREAD

**Issue by brandonwillard** [aesara-devs/aesara#748]
Now that we have a framework for static shape information, we need to update our use of `Rebroadcast`.

For some background, `Rebroadcast`s currently serve two functional roles:

1.  Generate a new `TensorVariable` with a specific `TensorType`/broadcast pattern from an existing `TensorVariable` (at the graph level).
    This is primarily used during the rewriting process to force the output type of a replacement to match the type it's replacing.
2.  Assert/check the broadcast pattern (at run-time).

The latter isn't necessary, especially since rewrites principally operate on the assumption that the replacement type is equivalent to the type being replaced. Without that assumption, rewrites wouldn't be reasonable in the first place. Run-time static shape information checking is effectively a leak of static analysis concerns into non-static analysis domains.

If a user wants to add their own asserts, fine, but static analysis and graph optimization shouldn't be adding asserts about their own fundamental operational assumptions into the output/results of a graph&#x2013;at least not unless debugging is enabled.

In other words, `Rebroadcast`s should not appear in the final graph output, and, as a result, we need a rewrite that replaces them with clones of their input nodes, but ones that output the desired `TensorType`s.

Also, we need to replace the use of `Rebroadcast` in `TensorType.convert_variable` with an `Op` that can assign new static shape information.

For instance, this should succeed, but currently doesn't:

```python
v1 = TensorType(aesara.config.floatX, (None, None))()
t2 = TensorType(aesara.config.floatX, (1, 2))
res = t2.convert_variable(v1)
assert res.type.shape == (1, 2)
```

To accomplish this, we could extend `SpecifyShape` to include support for `None`'s, or repurpose (and rename) `Rebroadcast`.

**Comment by brandonwillard** [comment]
As I mentioned there, https://github.com/aesara-devs/aesara/pull/825 is a first step toward allowing us to replace 
https://github.com/aesara-devs/aesara/blob/80bfde1ee23f6239ae86cdaa42af9e383048b759/aesara/tensor/type.py#L327-L329
with something like 
```python
new_static_shape = tuple(s_v if s_v is not None else s for s_v, s in zip(var.type.shape, self.shape))
return specify_shape(var, new_static_shape)
```
The latter would return a `Variable` with a `TensorType` that has the combined static shape information of both `Type`s, and that's what we ultimately want.

That approach might also work without changes to `SpecifyShape`, especially if we change instances of `None` in `new_static_shape` with the results of `shape(var)` at the appropriate index.  We would definitely need to make sure a few useless versions of `SpecifyShape` with `Shape` inputs are removed during optimization, though.
